sky/provision/docker_utils.py

```diff
@@ -0,0 +1,347 @@
+# The default maximum number of bytes to allocate to the object store unless
+# overridden by the user.
+DEFAULT_OBJECT_STORE_MAX_MEMORY_BYTES = 200 * 10**9
+# The default proportion of available memory allocated to the object store
+DEFAULT_OBJECT_STORE_MEMORY_PROPORTION = 0.3
+
+
+def _check_helper(cname, template, docker_cmd):
+    return ' '.join([
+        docker_cmd, 'inspect', '-f', '"{{' + template + '}}"', cname, '||',
+        'true'
+    ])
+
+
+def check_docker_running_cmd(cname, docker_cmd):
+    return _check_helper(cname, '.State.Running', docker_cmd)
+
+
+def check_bind_mounts_cmd(cname, docker_cmd):
+    return _check_helper(cname, 'json .Mounts', docker_cmd)
+
+
+def check_docker_image(cname, docker_cmd):
+    return _check_helper(cname, '.Config.Image', docker_cmd)
+
+
+def docker_start_cmds(
+    image,
+    container_name,
+    user_options,
+    docker_cmd,
+):
+    """Generating docker start command without --rm.
+
+    The code is borrowed from `ray.autoscaler._private.docker`. The only
+    difference is that we don't use `--rm` flag here.
+    """
+
+    # for click, used in ray cli
+    env_vars = {'LC_ALL': 'C.UTF-8', 'LANG': 'C.UTF-8'}
+    env_flags = ' '.join(
+        ['-e {name}={val}'.format(name=k, val=v) for k, v in env_vars.items()])
+
+    user_options_str = ' '.join(user_options)
+    docker_run = [
+        docker_cmd,
+        'run',
+        # SkyPilot: Remove --rm flag to keep the container after `ray stop`
+        # is executed.
+        '--name {}'.format(container_name),
+        '-d',
+        '-it',
+        env_flags,
+        user_options_str,
+        '--net=host',
+        image,
+        'bash',
+    ]
+    return ' '.join(docker_run)
+
+
+def _with_interactive(cmd):
+    force_interactive = (
+        f'source ~/.bashrc; '
+        f'export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && ({cmd})')
+    return ['bash', '--login', '-c', '-i', shlex.quote(force_interactive)]
+
+
+# SkyPilot: New class to initialize docker containers on a remote node.
+# Adopted from ray.autoscaler._private.command_runner.DockerCommandRunner.
+class DockerInitializer:
+    """Initializer for docker containers on a remote node."""
+
+    def __init__(self, docker_config: dict,
+                 runner: 'command_runner.SSHCommandRunner', log_path: str):
+        self.docker_config = docker_config
+        self.container_name = docker_config['container_name']
+        self.runner = runner
+        self.home_dir = None
+        self.initialized = False
+        use_podman = docker_config.get('use_podman', False)
+        self.docker_cmd = 'podman' if use_podman else 'docker'
+        self.log_path = log_path
+
+    def _run(self, cmd, run_env='host') -> str:
+        if run_env == 'docker':
+            cmd = self._docker_expand_user(cmd, any_char=True)
+            cmd = ' '.join(_with_interactive(cmd))
+            cmd = (f'{self.docker_cmd} exec {self.container_name} /bin/bash -c '
```

Comment:
Why are we removing interactive shell that is default enabled by ray? (`-it`)

Reply:
IIUC you need to wrap your command in `self.runner.run()` too if you want to use `-it` here...
See https://github.com/ray-project/ray/blob/d2fc4823126927b2c54f89ec72fa3d24b442e6a3/python/ray/autoscaler/_private/command_runner.py#L362-L363

I guess its fine to keep current implementation since all functionality works well

Reply:
Hmm, we do wrap the command with the interactive things in `runner.run()`. I left a comment here for future investigation.